pkg/controller/v1beta1/inferenceservice/components/predictor.go

```diff
@@ -54,20 +54,82 @@ func NewPredictor(client client.Client, scheme *runtime.Scheme, inferenceService
-	if sourceURI := predictor.GetStorageUri(); sourceURI != nil {
-		annotations[constants.StorageInitializerSourceUriInternalAnnotationKey] = *sourceURI
-	}
+
 	hasInferenceLogging := addLoggerAnnotations(isvc.Spec.Predictor.Logger, annotations)
 	hasInferenceBatcher := addBatcherAnnotations(isvc.Spec.Predictor.Batcher, annotations)
+
 	// Add agent annotations so mutator will mount model agent to multi-model InferenceService's predictor
 	addAgentAnnotations(isvc, annotations, p.inferenceServiceConfig)
 
+	// Reconcile modelConfig
+	configMapReconciler := modelconfig.NewModelConfigReconciler(p.client, p.scheme)
+	if err := configMapReconciler.Reconcile(isvc); err != nil {
+		return err
+	}
+
+	// If Model is specified, prioritize using that. Otherwise, we will assume a framework object was specified.
+	if isvc.Spec.Predictor.Model != nil {
+		runtimes, err := isvc.Spec.Predictor.Model.GetSupportingRuntimes(p.client, isvc.Namespace)
+		if err != nil {
+			return err
+		}
+		if len(runtimes) == 0 {
+			return errors.New("No runtime found to support predictor")
+		}
+		// Get first supporting runtime.
+		sRuntime := runtimes[0]
+		if len(sRuntime.Containers) == 0 {
+			return errors.New("No container configuration found in selected serving runtime")
+		}
+		// Assume only one container is specified in runtime spec.
+		container, err = isvcutils.MergeRuntimeContainers(&sRuntime.Containers[0], &isvc.Spec.Predictor.Model.Container)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get runtime container")
+		}
+		if sourceURI := isvc.Spec.Predictor.Model.GetStorageUri(); sourceURI != nil {
+			annotations[constants.StorageInitializerSourceUriInternalAnnotationKey] = *sourceURI
+		}
+
+		mergedPodSpec, err := isvcutils.MergePodSpec(&sRuntime.ServingRuntimePodSpec, &isvc.Spec.Predictor.PodSpec)
+		if err != nil {
+			return errors.Wrapf(err, "failed to consolidate serving runtime PodSpecs")
+		}
+
+		// Other dependencies rely on the container to be a specific name.
+		container.Name = constants.InferenceServiceContainerName
+
+		podSpec = *mergedPodSpec
+		podSpec.Containers = []v1.Container{
+			*container,
+		}
+
+	} else {
+		predictor := isvc.Spec.Predictor.GetImplementation()
+		container = predictor.GetContainer(isvc.ObjectMeta, isvc.Spec.Predictor.GetExtensions(), p.inferenceServiceConfig)
+		// Knative does not support INIT containers or mounting, so we add annotations that trigger the
+		// StorageInitializer injector to mutate the underlying deployment to provision model data
+		if sourceURI := predictor.GetStorageUri(); sourceURI != nil {
+			annotations[constants.StorageInitializerSourceUriInternalAnnotationKey] = *sourceURI
+		}
+
+		if len(isvc.Spec.Predictor.PodSpec.Containers) == 0 {
+			isvc.Spec.Predictor.PodSpec.Containers = []v1.Container{
+				*container,
+			}
+		} else {
+			isvc.Spec.Predictor.PodSpec.Containers[0] = *container
+		}
+
+		podSpec = v1.PodSpec(isvc.Spec.Predictor.PodSpec)
```

Comment:
nit: could move this up a few lines so that the others can reference `podSpec` instead of `isvc.Spec.Predictor.PodSpec`?

Reply:
Sure, I will do that.